### PR TITLE
[EigenSolversApplication] Extend FEAST interface

### PR DESCRIPTION
**Pull request by qaumann**
This:

- Adds sorting functions to sort FEAST's result by largest/smallest real, imaginary part or absolute value. This is especially useful when solving a non-symmetric or complex-symmetric system, as FEAST does not sort the eigenvalues by itself. 
- Implements the different settings for real / complex number handling as suggested by @philbucher.
- Changes the behavior of the `eigen_solver_factory`: If no symmetry is defined, the problem is assumed to be symmetric. This should be the default case for most application cases in structural mechanics.

### CODE REVIEW

File: applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h [hunk 1]
```diff
@@ -78,20 +77,102 @@ namespace { // helpers namespace
...
+                [&rEigenvalues](size_t i1, size_t i2) {return rEigenvalues[i1] < rEigenvalues[i2];});
+        } else if( mOrder == "sm") {
+            std::stable_sort(idx.begin(), idx.end(),
+                [&rEigenvalues](size_t i1, size_t i2) {return std::abs(rEigenvalues[i1]) < std::abs(rEigenvalues[i2]);});
+        } else if( mOrder == "lr" || mOrder == "li" ) {
+            std::stable_sort(idx.begin(), idx.end(),
+                [&rEigenvalues](size_t i1, size_t i2) {return rEigenvalues[i1] > rEigenvalues[i2];});
+        } else if( mOrder == "lm") {
+            std::stable_sort(idx.begin(), idx.end(),
+                [&rEigenvalues](size_t i1, size_t i2) {return std::abs(rEigenvalues[i1]) > std::abs(rEigenvalues[i2]);});
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h]:
```suggestion
                [&rEigenvalues](std::size_t i1, std::size_t i2) {return std::abs(rEigenvalues[i1]) > std::abs(rEigenvalues[i2]);});
```

File: applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h [hunk 2]
```diff
@@ -78,20 +77,102 @@ namespace { // helpers namespace
...
+
+        if( mOrder == "sr" || mOrder == "si" ) {
+            std::stable_sort(idx.begin(), idx.end(),
+                [&rEigenvalues](size_t i1, size_t i2) {return rEigenvalues[i1] < rEigenvalues[i2];});
+        } else if( mOrder == "sm") {
+            std::stable_sort(idx.begin(), idx.end(),
+                [&rEigenvalues](size_t i1, size_t i2) {return std::abs(rEigenvalues[i1]) < std::abs(rEigenvalues[i2]);});
+        } else if( mOrder == "lr" || mOrder == "li" ) {
+            std::stable_sort(idx.begin(), idx.end(),
+                [&rEigenvalues](size_t i1, size_t i2) {return rEigenvalues[i1] > rEigenvalues[i2];});
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h]:
```suggestion
                [&rEigenvalues](std::size_t i1, std::size_t i2) {return rEigenvalues[i1] > rEigenvalues[i2];});
```

File: applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h [hunk 3]
```diff
@@ -78,20 +77,102 @@ namespace { // helpers namespace
...
+
+        std::vector<size_t> idx(rEigenvalues.size());
+        std::iota(idx.begin(), idx.end(), 0);
+
+        if( mOrder == "sr" || mOrder == "si" ) {
+            std::stable_sort(idx.begin(), idx.end(),
+                [&rEigenvalues](size_t i1, size_t i2) {return rEigenvalues[i1] < rEigenvalues[i2];});
+        } else if( mOrder == "sm") {
+            std::stable_sort(idx.begin(), idx.end(),
+                [&rEigenvalues](size_t i1, size_t i2) {return std::abs(rEigenvalues[i1]) < std::abs(rEigenvalues[i2]);});
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h]:
```suggestion
                [&rEigenvalues](std::size_t i1, std::size_t i2) {return std::abs(rEigenvalues[i1]) < std::abs(rEigenvalues[i2]);});
```

File: applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h [hunk 4]
```diff
@@ -78,20 +77,102 @@ namespace { // helpers namespace
...
+
+    template<typename TScalar>
+    struct SortingHelper
+    {
+        SortingHelper(std::string Order) : mOrder(Order) {};
+        // void Check();
+        template<typename MatrixType, typename VectorType>
+        void SortEigenvalues(VectorType&, MatrixType&);
+        private:
+            std::string mOrder;
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h:98]: maybe using an `enum class` would be more appropriate?

File: applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h [hunk 5]
```diff
@@ -78,20 +77,102 @@ namespace { // helpers namespace
...
+            std::stable_sort(idx.begin(), idx.end(),
+                [&rEigenvalues](size_t i1, size_t i2) {return std::abs(rEigenvalues[i1]) > std::abs(rEigenvalues[i2]);});
+        } else {
+            KRATOS_ERROR << "Invalid sort type. Allowed are sr, sm, si, lr, lm, li" << std::endl;
+        }
+
+        VectorType tmp_eigenvalues(rEigenvalues.size());
+        MatrixType tmp_eigenvectors(rEigenvectors.size1(), rEigenvectors.size2());
+
+        for( size_t i=0; i<rEigenvalues.size(); ++i ) {
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h]:
```suggestion
        for( std::size_t i=0; i<rEigenvalues.size(); ++i ) {
```

File: applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h [hunk 6]
```diff
@@ -78,20 +77,102 @@ namespace { // helpers namespace
...
+        }
+        rEigenvalues.swap(tmp_eigenvalues);
+        rEigenvectors.swap(tmp_eigenvectors);
+    }
+
+    template<> template<typename MatrixType, typename VectorType>
+    void SortingHelper<std::complex<double>>::SortEigenvalues(VectorType &rEigenvalues, MatrixType &rEigenvectors)
+    {
+        std::vector<size_t> idx(rEigenvalues.size());
+        std::iota(idx.begin(), idx.end(), 0);
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h]:
```suggestion
        std::vector<std::size_t> idx(rEigenvalues.size(), 0);
```

File: applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h [hunk 7]
```diff
@@ -78,20 +77,102 @@ namespace { // helpers namespace
...
+    };
+
+    template<> template<typename MatrixType, typename VectorType>
+    void SortingHelper<double>::SortEigenvalues(VectorType &rEigenvalues, MatrixType &rEigenvectors)
+    {
+        KRATOS_WARNING_IF("FeastEigensystemSolver", mOrder == "si") << "Attempting to sort by imaginary value. Falling back on \"sr\"" << std::endl;
+        KRATOS_WARNING_IF("FeastEigensystemSolver", mOrder == "li") << "Attempting to sort by imaginary value. Falling back on \"lr\"" << std::endl;
+
+        std::vector<size_t> idx(rEigenvalues.size());
+        std::iota(idx.begin(), idx.end(), 0);
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h]:
```suggestion
        std::vector<std::size_t> idx(rEigenvalues.size(), 0);
```

File: applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h [hunk 8]
```diff
@@ -78,20 +77,102 @@ namespace { // helpers namespace
...
+        rEigenvectors.swap(tmp_eigenvectors);
+    }
+
+    template<> template<typename MatrixType, typename VectorType>
+    void SortingHelper<std::complex<double>>::SortEigenvalues(VectorType &rEigenvalues, MatrixType &rEigenvectors)
+    {
+        std::vector<size_t> idx(rEigenvalues.size());
+        std::iota(idx.begin(), idx.end(), 0);
+
+        // const std::string t = mParam["sort_order"].GetString();
```
Comment (philbucher) [inline applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h]: leftover

File: kratos/python_scripts/eigen_solver_factory.py [hunk 9]
```diff
@@ -36,7 +40,11 @@ def ConstructSolver(settings):
         if kratos_utils.CheckIfApplicationsAvailable("EigenSolversApplication"):
             import KratosMultiphysics.EigenSolversApplication as EiSA
             if EiSA.HasFEAST():
-                if settings.Has("symmetric") and settings["symmetric"].GetBool():
+                if settings.Has("symmetric"):
```
Comment (philbucher) [inline kratos/python_scripts/eigen_solver_factory.py]: using the ternary operator would be cleaner IMO

**Review by philbucher — COMMENTED** [review]
some style comments

also better use `std::size_t` instead of just `size_t`

File: applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h (hunk 1 above)
Comment (qaumann) [reply applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h]: I don't know why I keep forgetting this..

File: applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h (hunk 4 above)
Comment (qaumann) [reply applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h:98]: I also thought about this, but then I need a converter from the json string to this enum. If you think this is clearer, I can do it of course.

File: kratos/python_scripts/eigen_solver_factory.py (hunk 9 above)
Comment (qaumann) [reply kratos/python_scripts/eigen_solver_factory.py]: done

File: applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h (hunk 4 above)
Comment (philbucher) [reply applications/EigenSolversApplication/custom_solvers/feast_eigensystem_solver.h:98]: ok

**Review by philbucher — APPROVED** [review]

**Comment by qaumann** [comment]
Thanks for reviewing. @oberbichler these should be my last changes in the application for a while, so you can go ahead and rename it. If you need support, I am happy to help.

**Pull request merged**